File: LibrarySource/Python/tastranslation/tastranslation/translate.py
# Copyright 2024 Uncover-F
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import requests

def get_endpoints():
    endpoints_url = 'https://raw.githubusercontent.com/Uncover-F/TAS/Uncover/.data/endpoints.json'
    try:
        response = requests.get(endpoints_url)
        response.raise_for_status()  # Raise exception for bad status
        return response.json()
    except requests.exceptions.RequestException as e:
        print(f'Error fetching endpoints: {e}')
        return []

def translate(text, source_lang='en', target_lang='fr'):
    endpoints = get_endpoints()
    if not endpoints:
        return {'error': 'Failed to retrieve endpoints'}
    
    params = {
        'text': text,
        'source_lang': source_lang,
        'target_lang': target_lang
    }
    
    for endpoint in endpoints:
        try:
            response = requests.get(endpoint, params=params)
            response.raise_for_status()  # Raise exception for bad status
            return response.json()
        except requests.exceptions.RequestException as e:
            print(f'Error at {endpoint}: {e}')
    
    return {'error': 'All endpoints failed'}

File: .data/js_libraries/translate.js
// Copyright 2024 Uncover-F

// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//     http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// JS Library for translating text using the TAS API
// https://github.com/Uncover-F/TAS

async function translateText(text, sourceLang, targetLang) {

    try {
        const endpointsResponse = await fetch(endpointsUrl);
        if (!endpointsResponse.ok) {
            throw new Error(`Error fetching endpoints: ${endpointsResponse.status} - ${endpointsResponse.statusText}`);
        }
        const endpointsData = await endpointsResponse.json();
        const endpoints = JSON.parse(endpointsData.contents);

        for (const endpoint of endpoints) {
            const apiUrl = `https://api.allorigins.win/get?url=${encodeURIComponent(`${endpoint}?text=${encodeURIComponent(text)}&source_lang=${sourceLang}&target_lang=${targetLang}`)}`;
            try {
                const response = await fetch(apiUrl);
                if (response.ok) {
                    const data = await response.json();
                    const parsedContents = JSON.parse(data.contents);
                    return parsedContents.response.translated_text;
                } else {
                    console.error(`Error at ${endpoint}: ${response.status} - ${response.statusText}`);
                }
            } catch (error) {
                console.error(`Request exception at ${endpoint}:`, error);
            }
        }
        throw new Error('All endpoints failed.');
    } catch (error) {
        console.error('Error:', error);
        throw error;
    }
}

export default translateText;
